singularity-eos/eos/eos_vinet.hpp

```diff
@@ -0,0 +1,366 @@
+//------------------------------------------------------------------------------
+// © 2021-2022. Triad National Security, LLC. All rights reserved.  This
+// program was produced under U.S. Government contract 89233218CNA000001
+// for Los Alamos National Laboratory (LANL), which is operated by Triad
+// National Security, LLC for the U.S.  Department of Energy/National
+// Nuclear Security Administration. All rights in the program are
+// reserved by Triad National Security, LLC, and the U.S. Department of
+// Energy/National Nuclear Security Administration. The Government is
+// granted for itself and others acting on its behalf a nonexclusive,
+// paid-up, irrevocable worldwide license in this material to reproduce,
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef _SINGULARITY_EOS_EOS_VINET_HPP_
+#define _SINGULARITY_EOS_EOS_VINET_HPP_
+
+#include <cmath>
+#include <cstdio>
+
+#include <limits>
+
+#include <singularity-eos/base/constants.hpp>
+#include <singularity-eos/base/math_utils.hpp>
+#include <singularity-eos/base/root-finding-1d/root_finding.hpp>
+#include <singularity-eos/eos/eos_base.hpp>
+
+namespace singularity {
+
+using namespace eos_base;
+
+// COMMENT: This is a complete, thermodynamically consistent Vinet
+// EOS created at Sandia National Laboratories by AEM.
+// J. Appl. Phys. 119, 015904 (2016)
+// It is particularly suited for high-pressure applications.
+class Vinet : public EosBase<Vinet> {
+ public:
+  Vinet() = default;
+  // Constructor 
+  PORTABLE_INLINE_FUNCTION
+  Vinet(const Real rho0, const Real T0, const Real B0, const Real BP0, const Real A0, 
+	    const Real Cv0, const Real E0, const Real S0, const Real *expconsts)
+      : _rho0(rho0), _T0(T0), _B0(B0), _BP0(BP0), _A0(A0), _Cv0(Cv0), _E0(E0), _S0(S0), 
+	_expconsts(expconsts) {
+      CheckVinet();
+  }
```

Comment:
Is there a way to structure this so that either it's explicit that `expconsts` needs 39 entries or in a way where all 39 don't need to be specified and a number of specified values can be given instead?

Reply:
It is assumed the host code sets and initialize (to 0.0) a constant array of length 39. Then they read in the parameters the user have given in the appropriate box, in any way they want, and send it in to singularity. I have done this in many different ways in different host-codes and found that having a fixed size array is the best way for a host code developer to get the correct thing in.

Reply:
Responding to your above comment,
> If you have a way to send the full array and not a pointer to the Vinet constructor, that would be best. I could not find a way. In any case, the developer in the host code need to know what size array they should send in so it we cannot rely on them reading the instructions enough to find the size out we need to add a MAXVINETPARAMETERS that they can ask for. Can I check on the size of the array the pointer is pointing at? Then I could introduce a check that this is OK as well.

Without using standard library functions (because I'm not certain what would work on GPUs), we may not have a way of figuring out the size of the input parameter array, you're right.

>It is assumed the host code sets and initialize (to 0.0) a constant array of length 39. Then they read in the parameters the user have given in the appropriate box, in any way they want, and send it in to singularity. I have done this in many different ways in different host-codes and found that having a fixed size array is the best way for a host code developer to get the correct thing in.

Let me see if I can organize my thoughts better, but it all boils down to why 39 parameters? Why not more or why not less, and would a user (either through a host code or some other means) want to specify only the leading terms if that's good enough?

As long as there are no fundamental restrictions on the number of terms in the series expansion, you could structure the code to take a specified number of terms and not be dependent on there being exactly 39 terms. The user would pass a pointer to the coefficients and tell you how many are in the coefficient matrix and the Vinet EOS could figure out the correct counters for the loops lengths and number of provided parameters.

Reply:
It also seems to me that not all 39 terms are used right? The zero terms aren't even looped over. It doesn't seem to me a fixed size array is needed. I suggest `std::vector`.

Reply:
Yeah you're right, I forgot this is in the constructor so standard library functions are fine. We could also template on the input and just require some behavior like `std::vector`